include/nbl/builtin/hlsl/emulated/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,622 @@
+                uint64_t rhsNormMantissa = ieee754::extractNormalizeMantissa(rhsData);
+
+                const int expDiff = lhsBiasedExp - rhsBiasedExp;
+
+                const int exp = max(lhsBiasedExp, rhsBiasedExp) - ieee754::traits<float64_t>::exponentBias;
+                const uint32_t shiftAmount = abs(expDiff);
+
+                if (expDiff < 0)
+                {
+                    // so lhsNormMantissa always holds mantissa of number with greater exponent
+                    swap<uint64_t>(lhsNormMantissa, rhsNormMantissa);
+                    swap<uint64_t>(lhsSign, rhsSign);
+                }
+
+                rhsNormMantissa >>= shiftAmount;
+
+                uint64_t resultMantissa;
+                if (lhsSign != rhsSign)
+                {
+                    int64_t mantissaDiff = lhsNormMantissa - rhsNormMantissa;
+                    if (mantissaDiff < 0)
+                    {
+                        swap<uint64_t>(lhsNormMantissa, rhsNormMantissa);
+                        swap<uint64_t>(lhsSign, rhsSign);
+                    }
+
+                    lhsNormMantissa <<= 10;
+                    rhsNormMantissa <<= 10;
+                    resultMantissa = uint64_t(int64_t(lhsNormMantissa) - int64_t(rhsNormMantissa));
+                    resultMantissa >>= 10;
+                }
+                else
+                {
+                    resultMantissa = lhsNormMantissa + rhsNormMantissa;
+                }
+
+                uint64_t resultBiasedExp = uint64_t(exp) + ieee754::traits<float64_t>::exponentBias;
+
+                if (resultMantissa & 1ull << 53)
+                {
+                    ++resultBiasedExp;
+                    resultMantissa >>= 1;
+                }
+
+                while (resultMantissa < (1ull << 52))
+                {
+                    --resultBiasedExp;
+                    resultMantissa <<= 1;
+                }
+
+                resultMantissa &= ieee754::traits<float64_t>::mantissaMask;
+                uint64_t output = impl::assembleFloat64(lhsSign, resultBiasedExp << ieee754::traits<float64_t>::mantissaBitCnt, resultMantissa);
+                return bit_cast<this_t>(output);
+            }
+
+            // not implemented
+            if (!FlushDenormToZero)
+                return bit_cast<this_t>(0xdeadbeefbadcaffeull);
+        }
+
+        emulated_float64_t operator+(float rhs)
+        {
+            return bit_cast<this_t>(data) + create(rhs);
+        }
+
+        emulated_float64_t operator-(emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
+            emulated_float64_t lhs = bit_cast<this_t>(data);
+            emulated_float64_t rhsFlipped = rhs.flipSign();
+            
+            return lhs + rhsFlipped;
+        }
+
+        emulated_float64_t operator-(float rhs) NBL_CONST_MEMBER_FUNC
+        {
+            return bit_cast<this_t>(data) - create(rhs);
+        }
```

Comment:
using `bit_cast` to promote a lower bit-depth float is wrong, should be `_static_cast` (or `create` if you can't do `_static_cast` due to circular refs)

P.S. `bit_cast` should static_assert/crash due to `sizeof(T)!=sizeof(U)` if you actually tried to use these operators

Reply:
> using bit_cast to promote a lower bit-depth float is wrong
but i never do that, `data` is `uint64_t`

Reply:
also `bit_cast` template specializations are defined only for `float64_t` and `uint64_t`